superdesk/commands/async_cli.py

```diff
@@ -0,0 +1,107 @@
+from functools import wraps
+from typing import Any, Callable, Optional, Tuple, cast
+from asgiref.sync import async_to_sync
+
+from ..flask import Blueprint, AppGroup, Flask
+
+
+class AsyncAppGroup(AppGroup):
+    """
+    An extension of Quart's AppGroup to support registration of asynchronous command handlers.
+
+    This class provides a mechanism to register asynchronous functions as command line commands,
+    which are automatically handled synchronously using `asgiref.sync.async_to_sync` conversion.
+    """
+
+    app: Flask
+
+    def register_async_command(
+        self, name: Optional[str] = None, with_appcontext=False, **click_kwargs: dict[str, Any]
+    ) -> Callable[..., Any]:
+        """
+        A decorator to register an asynchronous command within the Quart CLI app group.
+        If your command needs to use the app context, pass `with_appcontext=True` to ensure it will be run
+        wrapped with current app context. Otherwise getting current app context will fail in asynchronous
+        command contexts due to execution in different threads or outside of the normal request context lifecycle.
+
+        Args:
+            name (str, optional): The name of the command. Defaults to the function's name if None.
+            with_appcontext (bool): If True, it will wrap the command function in the current_app's context.
+                                    Note that `set_current_app` method has to be called first.
+            **click_kwargs: Additional keyword arguments that are passed to the `AppGroup.command` decorator.
+
+        Example:
+        Use the decorator to register a command that requires app context:
+
+        .. code-block:: python
+
+            from superdesk.commands import cli
+
+            @cli.register_async_command("example-command", with_appcontext=True)
+            async def my_command():
+                app = get_current_app()
+                with app.app_context():
+                    # your command logic here
+        """
+
+        def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
+            @wraps(func)
+            # This wrapper will convert the async function to a sync function
+            # using async_to_sync when the command is actually called.
+            def sync_wrapper(*args, **kwargs) -> Any:
+                if with_appcontext:
+                    return async_to_sync(self.run_with_appcontext)(func, *args, **kwargs)
```

Comment:
We'll see how it goes using `async_to_sync`. My experience in the past, when we were on Flask3, is that this doesn't play nicely with PyMongo / Elasticsearch clients. Issues such as the connections to the DBs having a different event loop associated then the one currently being executed (assuming because it's not thread-safe, and not using the same context).

Reply:
Yeah, that was one of my concerns with the run_async_to_sync that I added on newsroom-core. I'm hoping that wrapping the command in the app context before executing it with async_to_sync would be enough to prevent some of those issues. My previous approach was instead, passing the app instance object directly to the command, and if the command needs the app then use that one instead of getting it from the `get_current_app` function, but still there is the risk that the connections are generated in a separate event loop before running the command. Anyways, we'll see how it goes.